src/javascript/ContentEditor/utils/useReorderList.js

```diff
@@ -0,0 +1,165 @@
+import {useDrag, useDrop} from 'react-dnd';
+import {useCallback, useRef, useState} from 'react';
+import {uniqueId} from 'lodash';
+
+export function onListIndexReorder(list, dragIndex, hoverIndex) {
+    const newList = [...list]; // Create a copy of the original list
+    const [draggedItem] = newList.splice(dragIndex, 1); // Remove the dragged item
+    newList.splice(hoverIndex, 0, draggedItem); // Insert the dragged item at the hover position
+    return newList;
+}
+
+function makeDragItems(items) {
+    return items.map((item, index) => ({
+        item,
+        index,
+        id: uniqueId()
+    }));
+}
+
+function updateDragItems(items, reorderedItems) {
+    return items.map((item, index) => {
+        return {...reorderedItems[index], item, index};
+    });
+}
+
+/**
+ * Custom hook to manage the reordering of a list of items.
+ * It initializes the list with drag items and provides functions to handle reordering and resetting the list.
+ * @param {array} items    array of items to be reordered
+ * @returns {object}    object containing the reordered items, handleReorder function, and reset function
+ */
+export function useReorderList(items) {
+    // Note the use of serialization to detect changes in the items array
+    // this is necessary to avoid unnecessary re-renders when the items array is the same, because formik values are not stable
+    const itemsRef = useRef({items, serialized: JSON.stringify(items)});
+    const currentItemsSerialized = JSON.stringify(items);
+
+    const [reorderedItems, setReorderedItems] = useState(makeDragItems(items));
+
+    const reset = useCallback(() => {
+        setReorderedItems(makeDragItems(items));
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [currentItemsSerialized]);
+
+    const handleReorder = useCallback((dragIndex, hoverIndex) => {
+        setReorderedItems(prevItems =>
+            onListIndexReorder(prevItems, dragIndex, hoverIndex).map((item, index) => ({...item, index}))
+        );
+    }, []);
+
+    if (currentItemsSerialized !== itemsRef.current.serialized) {
+        /*
+        The significance of this check is to have latest values available immediately, without an extra render
+         */
+        let newReorderedItems;
+        if (itemsRef.current.items.length === items.length) {
+            newReorderedItems = updateDragItems(items, reorderedItems);
+        } else {
+            newReorderedItems = makeDragItems(items);
+        }
+
+        setReorderedItems(newReorderedItems);
+        itemsRef.current = {items, serialized: JSON.stringify(items)};
+        return {reorderedItems: newReorderedItems, handleReorder, reset};
```

Comment:
is return here necessary ? wouldn't the set call on line 62 trigger the return call on line 67 with new value ?